Sherlock/sherlock.h

```diff
@@ -417,178 +398,192 @@ class StreamImpl {
-      CURRENT_THROW(StreamInGracefulShutdownException());
-    }
+    return SubscriberScope<F, TYPE_SUBSCRIBED_TO>(impl_, subscriber, begin_idx, done_callback);
+  }
+
+  // Generates a random HTTP subscription.
+  static std::string GenerateRandomHTTPSubscriptionID() {
+    return current::SHA256("sherlock_http_subscription_" +
+                           current::ToString(current::random::CSRandomUInt64(0ull, ~0ull)));
   }
 
   // Sherlock handler for serving stream data via HTTP (see `pubsub.h` for details).
   template <class J>
-  void ServeDataViaHTTP(Request r) {
-    try {
-      // Prevent `own_data_` from being destroyed between the entry into this function
-      // and the call to the construction of `PubSubHTTPEndpoint`.
-      //
-      // Granted, an overkill, as whoever could call `ServeDataViaHTTP` from another thread should have
-      // its own `ScopeOwnedBySomeoneElse<>` copy of the stream object. But err on the safe side. -- D.K.
-      ScopeOwnedBySomeoneElse<stream_data_t> scoped_data(own_data_, []() {});
-      stream_data_t& data = *scoped_data;
-
-      auto request_params = ParsePubSubHTTPRequest(r);  // Mutable as `tail` may change. -- D.K.
-
-      if (request_params.terminate_requested) {
-        typename stream_data_t::http_subscriptions_t::iterator it;
-        {
-          // NOTE: This is not thread-safe!!!
-          // The iterator may get invalidated in between the next few lines.
-          // However, during the call to `= nullptr` the mutex will be locked again from within the
-          // thread terminating callback. Need to clean this up. TODO(dkorolev).
-          std::lock_guard<std::mutex> lock(data.http_subscriptions_mutex);
-          it = data.http_subscriptions.find(request_params.terminate_id);
-        }
-        if (it != data.http_subscriptions.end()) {
-          // Subscription found. Delete the scope, triggering the thread to shut down.
-          // TODO(dkorolev): This should certainly not happen synchronously.
-          it->second.first = nullptr;
-          // Subscription terminated.
-          r("", HTTPResponseCode.OK);
-        } else {
-          // Subscription not found.
-          r("", HTTPResponseCode.NotFound);
-        }
-        return;
-      }
+  void ServeDataViaHTTP(Request r) const {
+    if (r.method != "GET" && r.method != "HEAD") {
+      r(current::net::DefaultMethodNotAllowedMessage(), HTTPResponseCode.MethodNotAllowed);
+      return;
+    }
 
-      // Unsupported HTTP method.
-      if (r.method != "GET" && r.method != "HEAD") {
-        r(current::net::DefaultMethodNotAllowedMessage(), HTTPResponseCode.MethodNotAllowed);
-        return;
-      }
+    // First, ensure we're not in the destruction mode, and atomically claim we should not be
+    // without destructing this HTTP subscription, assuming we do follow up with creating one.
+    const Borrowed<impl_t> borrowed_impl(impl_);
+    if (!borrowed_impl) {
```

Comment:
`IsValid` would do (and so would just `if (impl_)`, but the subscriber should borrow the stream data to work with it and to make sure it doesn't disappear in the meantime.